futures-core/src/task/mod.rs

```diff
@@ -1,35 +1,142 @@
+//! task-local data generated by `task_local!`.
+//!
+//! Note that libraries typically should not manage tasks themselves, but rather
+//! leave that to event loops and other "executors" (see the `executor` module),
+//! or by using the `wait` method to create and execute a task directly on the
+//! current thread.
+//!
+//! More information about the task model can be found [online at tokio.rs].
+//!
+//! [online at tokio.rs]: https://tokio.rs/docs/going-deeper-futures/futures-model/
+//!
+//! ## Functions
+//!
+//! There is an important bare function in this module: `current`. The
+//! `current` function returns a handle to the currently running task, panicking
+//! if one isn't present. This handle is then used to later notify the task that
+//! it's ready to make progress through the `Task::notify` method.
+
 use core::fmt;
 use core::marker::PhantomData;
 
-use {Poll, Future, Stream};
-
 mod atomic_task;
 pub use self::atomic_task::AtomicTask;
 
-mod core;
+/// The context for the currently running task.
+pub struct Context<'a> {
+    handle_factory: &'a (Fn() -> NotifyHandle + 'a),
+    id: usize,
+    map: &'a LocalMap,
+}
 
-#[cfg(feature = "std")]
-mod std;
-#[cfg(feature = "std")]
-pub use self::std::*;
-#[cfg(not(feature = "std"))]
-pub use self::core::*;
+impl<'a> Context<'a> {
+    /// TODO: dox
+    pub fn new(
+        map: &'a LocalMap,
+        id: usize,
+        handle_factory: &'a (Fn() -> NotifyHandle + 'a),
+    )
+        -> Context<'a>
+    {
+        Context {
+            handle_factory,
+            id,
+            map,
+        }
+    }
 
-/// The context for the currently running task.
-#[derive(Debug)]
-pub struct Context;
+    /// Returns a handle to the current task to call `notify` at a later date.
+    ///
+    /// The returned handle implements the `Send` and `'static` bounds and may
+    /// also be cheaply cloned. This is useful for squirreling away the handle
+    /// into a location which is then later signaled that a future can make
+    /// progress.
+    ///
+    /// Implementations of the `Future` trait typically use this function if
+    /// they would otherwise perform a blocking operation. When something isn't
+    /// ready yet, this `current` function is called to acquire a handle to the
+    /// current task, and then the future arranges it such that when the
+    /// blocking operation otherwise finishes (perhaps in the background) it
+    /// will `notify` the returned handle.
+    ///
+    /// It's sometimes necessary to pass extra information to the task when
+    /// unparking it, so that the task knows something about *why* it was woken.
+    /// See the `FutureQueue` documentation for details on how to do this.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if a task is not currently being executed.
+    /// That is, this method can be dangerous to call outside of an
+    /// implementation of `poll`.
+    pub fn waker(&mut self) -> Task {
```

Comment:
TODO: make this return a `Waker` rather than a `Task`.

Reply:
Sure yeah, I can do that in this PR as well.